FormalSystems/Computability/Loop.lean

```diff
@@ -0,0 +1,200 @@
+import Mathlib.Data.Finite.Defs
+import Mathlib.Data.Finset.Fold
+import Mathlib.Data.Finset.Prod
+
+import FormalSystems.Preliminaries.Fold
+
+structure Variable where
+  index: Nat
+  deriving DecidableEq
+
+theorem Variable.eq (v₁ v₂: Variable): v₁ = v₂ ↔ v₁.index = v₂.index := by
+  constructor
+  . intro; congr
+  . intro h
+    cases v₁; cases v₂
+    dsimp at h; rw [h]
+
+inductive LoopProgram
+| add (x y: Variable) (n: Nat)
+| sub (x y: Variable) (n: Nat)
+| concat (p₁ p₂: LoopProgram)
+| loop (x: Variable) (p: LoopProgram)
+deriving DecidableEq
+
+def LoopProgram.toFunction (p: LoopProgram):
+  Nat → Nat :=
+  -- Todo: define how to evaluate `p`, with x₀ bound to the input
+  -- for example one could use a state monad with an `AssocList`
+  sorry
```

Comment:
Do you still plan to implement this?
Otherwise, you can leave this for me as an exercise :)

Reply:
I haven't worked on this yet. So go ahead if you feel inspired :)

Reply:
Will do!
I merge this now and will create a separate PR for the implementation.